synir/src/data_structures/pauli_polynomial.rs

```diff
@@ -62,6 +61,38 @@ impl PauliPolynomial {
     pub fn angle(&self, i: usize) -> Angle {
         self.angles[i]
     }
+
+    pub fn commutes_with(&self, other: &PauliPolynomial) -> bool {
+        let size = self.size();
+        assert_eq!(size, other.size());
+
+        let self_length = self.length();
+        let other_length = other.length();
+
+        for index_1 in 0..self_length {
+            let mut pauli_string = Vec::with_capacity(size);
+            for q1 in 0..size {
+                pauli_string.push(self.chain(q1).pauli(index_1));
+            }
+            for index_2 in 0..other_length {
+                let mut other_pauli_string = Vec::with_capacity(size);
+                for q2 in 0..size {
+                    other_pauli_string.push(other.chain(q2).pauli(index_2));
+                }
+                let mut commutes = true;
+                for (p1, p2) in zip(&pauli_string, &other_pauli_string) {
```

Comment:
Since `other_pauli_string` is only iterated once, an iterator is sufficient, no need to collect into a Vec.
```rust
let other_pauli_string = (0..size).map(|q2| other.chain(q2).pauli(index_2));
```
For the outer `pauli_string` it's a different story; it's used multiple times and collecting it once could be cheaper since the access is non-sequential.